go/toolchains.rst

```diff
@@ -18,22 +27,568 @@ customize the behavior of the core_ go rules.
 Design
 ------
 
-**TODO**: More information
+The go toolchain consists of two main layers, `the sdk`_ and `the toolchain`_.
+
+The SDK
+~~~~~~~
+
+At the bottom is the go sdk. This is the same thing you would get if you go to the main
+`go website`_ and download a `binary distribution`_.
+
+The go_sdk_ rule is responsible for downloading these, and adding just enough of a build file
+to expose the contents to Bazel. It currently also builds the cross compiled standard libraries
+for specific combinations, although we hope to make that an on demand step in the future.
+
+SDK's are specific to the host they are running on, and the version of go they want to use,
+but not the target they compile for. The go SDK is naturally `cross compiling`_.
+
+The go rules already adds a go_sdk_ rule for all the host and go version pairs that are shipped
+on the main go language website, so there should be no need to declare one of these unless you
+need a `forked version of go`_, however you may want to `control the version`_ or use the
+`installed sdk`_.
+
+The toolchain
+~~~~~~~~~~~~~
+
+Declaration
+^^^^^^^^^^^
+
+Toolchains are declared using the go_toolchain_ macro. This actually registers two Bazel
+toolchains, the main go toolchain, and a special bootstrap toolchain. The bootstrap toolchain
+is needed because the full toolchain includes tools that are compiled on demand and written in
+go, so we need a special cut down version of the toolchain to build those tools.
+
+Toolchains are pre-declared for all the known combinations of host, target and sdk, and the names
+are a predictable
+go\ **version**\ _\ **host**
+for host toolchains and
+go\ **version**\ _\ **host**\ _cross\_\ **target**
+for cross compilation toolchains. So for instance if the rules_go repository is loaded with
+it's default name, the following toolchain labels (along with many others) will be available
+
+.. code::
+
+  @io_bazel_rules_go//go/toolchain:1.9.0_linux_amd64
+  @io_bazel_rules_go//go/toolchain:1.9.0_linux_amd64-bootstrap
+  @io_bazel_rules_go//go/toolchain:1.9.0_linux_amd64_cross_windows_amd64
+
+The toolchains are not usable until you register_ them.
+
+Registration
+^^^^^^^^^^^^
+
+Normally you would just call go_register_toolchains_ from your WORKSPACE to register all the
+pre-declared toolchains, and allow normal selection logic to pick the right one.
+It is fine to add more toolchains to the set available if you like, and because the normal
```

Comment:
Couple small fixes:

It is fine to add more toolchains to the available set if you like. Because the normal toolchain matching mechanism prefers the first declared match, you can also override individual toolchains by declaring and registering toolchains with the same constraints *before* calling `go_register_toolchains`.